Web Inspector: Add emulation toggles for prefers-reduced-motion and prefers-contrast

rcaliman-apple opened the pull request:
#### e7a763ba3abd9f265af5c63531dd69a451937085
<pre>
Web Inspector: Add emulation toggles for prefers-reduced-motion and prefers-contrast
<a href="https://bugs.webkit.org/show_bug.cgi?id=246882">https://bugs.webkit.org/show_bug.cgi?id=246882</a>

Reviewed by Patrick Angle.

Adds controls in Web Inspector to override preferences
for reduced motion and increased contrast for the inspected page.

These correspond to CSS media features `prefers-reduced-motion: reduce | no-preference`
and `prefers-contrast: more | no-preference`.

The UI in Web Inspector starts off with preference values set to `System`
meaning there is no override applied. A user can pick `On` or `Off` to
force a value for the preference on the inspected page for the duration of
the Web Inspector session.

When a preference is overriden in Web Inspector, changes after the fact to
corresponding settings in System Settings &gt; Accessibility &gt; Display
have no effect on the inspected page until Web Inspector is closed or
the preference values are returned to `System`.

* LayoutTests/inspector/page/overrideUserPreference-expected.txt: Added.
* LayoutTests/inspector/page/overrideUserPreference.html: Added.
* Source/JavaScriptCore/inspector/protocol/Page.json:
* Source/WebCore/inspector/InspectorInstrumentation.cpp:
(WebCore::InspectorInstrumentation::accessibilitySettingsDidChangeImpl):
* Source/WebCore/inspector/InspectorInstrumentation.h:
(WebCore::InspectorInstrumentation::accessibilitySettingsDidChange):
* Source/WebCore/inspector/agents/InspectorPageAgent.cpp:
(WebCore::InspectorPageAgent::enable):
(WebCore::InspectorPageAgent::disable):
(WebCore::InspectorPageAgent::overrideUserPreference):
(WebCore::InspectorPageAgent::overridePrefersReducedMotion):
(WebCore::InspectorPageAgent::overridePrefersContrast):
(WebCore::InspectorPageAgent::accessibilitySettingsDidChange):
(WebCore::InspectorPageAgent::defaultUserPreferencesDidChange):
* Source/WebCore/inspector/agents/InspectorPageAgent.h:
* Source/WebCore/page/Page.cpp:
(WebCore::Page::accessibilitySettingsDidChange):
* Source/WebInspectorUI/Localizations/en.lproj/localizedStrings.js:
* Source/WebInspectorUI/UserInterface/Controllers/CSSManager.js:
(WI.CSSManager):
(WI.CSSManager.prototype.get overridenUserPreferences):
(WI.CSSManager.prototype.get defaultUserPreferences):
(WI.CSSManager.prototype.overrideUserPreference):
(WI.CSSManager.prototype.defaultUserPreferencesDidChange):
* Source/WebInspectorUI/UserInterface/Images/AppearanceOverride.svg: Added.
* Source/WebInspectorUI/UserInterface/Main.html:
* Source/WebInspectorUI/UserInterface/Protocol/PageObserver.js:
(WI.PageObserver.prototype.defaultUserPreferencesDidChange):
* Source/WebInspectorUI/UserInterface/Views/DOMTreeContentView.js:
(WI.DOMTreeContentView):
(WI.DOMTreeContentView.prototype.get navigationItems):
(WI.DOMTreeContentView.prototype._handleOverrideUserPreferencesButtonClicked):
(WI.DOMTreeContentView.prototype._overridenUserPreferencesDidChange):
(WI.DOMTreeContentView.prototype.didDismissPopover):
* Source/WebInspectorUI/UserInterface/Views/OverrideUserPreferencesPopover.css: Added.
(.popover .user-preferences-content):
(.popover .user-preferences-content &gt; h1):
(.popover .user-preferences-content &gt; label):
(.popover .user-preferences-content &gt; select):
* Source/WebInspectorUI/UserInterface/Views/OverrideUserPreferencesPopover.js: Added.
(WI.OverrideUserPreferencesPopover):
(WI.OverrideUserPreferencesPopover.prototype.show):
(WI.OverrideUserPreferencesPopover.prototype.dismiss):
(WI.OverrideUserPreferencesPopover.prototype._presentOverTargetElement):
(WI.OverrideUserPreferencesPopover.prototype._createSelectElement):
(WI.OverrideUserPreferencesPopover.prototype._createContentElement):
(WI.OverrideUserPreferencesPopover.prototype._userPreferenceValueLabel):
(WI.OverrideUserPreferencesPopover.prototype._defaultUserPreferencesDidChange):

Canonical link: <a href="https://commits.webkit.org/257197@main">https://commits.webkit.org/257197@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/239471dea5ae6a1340bf46e9fb38f531d9238b1e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/98158 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/7371 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/31303 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/107618 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/168446 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/102100 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/7864 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/84754 "Built successfully") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/90747 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/104207 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/103810 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/5907 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/89488 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/90747 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/87789 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/20920 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/36/builds/90747 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/88997 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/1335 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/22423 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/84728 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/1294 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/44878 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/28630 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4954 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/6168 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/41845 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/87568 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/2629 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/19638 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

dcrousso commented on Source/WebCore/inspector/agents/InspectorPageAgent.h:
```diff
@@ -158,6 +160,7 @@ class InspectorPageAgent final : public InspectorAgentBase, public Inspector::Pa
 
     Ref<Inspector::Protocol::Page::Frame> buildObjectForFrame(Frame*);
     Ref<Inspector::Protocol::Page::FrameResourceTree> buildObjectForFrameTree(Frame*);
+    Ref<Inspector::Protocol::Page::AccessibilitySetting> buildObjectForAccessibilitySetting(Inspector::Protocol::Page::AccessibilitySetting::Name name, bool settingEnabledOnSystem, ForcedAccessibilityValue forcedAccessibilityValue);
```
I think this should be `const Inspector::Protocol::Page::AccessibilitySetting::Name&`.

Style: the `name` and `forcedAccessibilityValue` parameter names are redundant (i.e. the type is clear enough)

dcrousso commented on Source/JavaScriptCore/inspector/protocol/Page.json:
```diff
@@ -385,6 +397,14 @@
+            "description": "Fired when a page's accessibility settings change, either by the system or forced.",
+            "targetTypes": ["page"],
+            "parameters": [
+                { "name": "accessibilitySettings", "type": "array", "items": { "$ref": "AccessibilitySetting"}, "description": "List of AccessibilitySetting instances" }
```
Style: missing space before `}`

NIT: can we have a better description?

dcrousso commented on Source/WebCore/inspector/agents/InspectorPageAgent.cpp:
```diff
@@ -356,6 +357,8 @@ Protocol::ErrorStringOr<void> InspectorPageAgent::enable()
 #if ENABLE(DARK_MODE_CSS) || HAVE(OS_DARK_MODE_SUPPORT)
     defaultAppearanceDidChange(m_inspectedPage.defaultUseDarkAppearance());
 #endif
+    // Let the Web Inspector frontend know about the state of system preferences.
```
i'd remove this comment as doesn't really add anything that isn't already derivable from the code

dcrousso commented on Source/JavaScriptCore/inspector/protocol/Page.json:
```diff
@@ -21,7 +21,19 @@
+            "properties": [
+                { "name": "name", "type": "string", "enum": ["prefersReducedMotion", "prefersContrast"], "description": "Name of accessibility setting" },
+                { "name": "enabled", "type": "boolean", "description": "Whether the setting is currently enabled, regardless if applied by the system or forced." },
+                { "name": "forced", "type": "boolean", "description": "Whether the forced state of the setting is the opposite of the current system value, whichever it may be." }
```
I'm not really sure I see the need for a separate `forced` flag.  Anything from Web Inspector is kinda always "forced" in a sense.

dcrousso commented on Source/WebCore/inspector/agents/InspectorPageAgent.cpp:
```diff
@@ -457,6 +462,16 @@ Protocol::ErrorStringOr<void> InspectorPageAgent::overrideSetting(Protocol::Page
         return { };
 
+    case Protocol::Page::Setting::ForcedPrefersReducedMotion:
+        inspectedPageSettings.setForcedPrefersReducedMotionAccessibilityValue(value == true ? ForcedAccessibilityValue::On : ForcedAccessibilityValue::Off);
```
Are these used anywhere else?  Do we need to worry about Web Inspector clobbering over the state and not restoring it properly afterwards?

Also, this would not match the expected use of the protocol.  If a developer calls `Page.overrideSetting("PrefersReducedMotion")` then they'd expect that any override Web Inspector sets would be removed, not that it would resolve to `Off`.
```
inspectedPageSettings.setForcedPrefersReducedMotionAccessibilityValue(value ? (*value ? ForcedAccessibilityValue::On : ForcedAccessibilityValue::Off) : ForcedAccessibilityValueSystem);
```

dcrousso commented on Source/JavaScriptCore/inspector/protocol/Page.json:
```diff
@@ -21,7 +21,19 @@
-                "WebSecurityEnabled"
+                "WebSecurityEnabled",
+                "ForcedPrefersReducedMotion",
+                "ForcedPrefersContrast"
```
I realize that `InternalSettings` include `Forced` in the name, but I don't think we need to include that in the inspector protocol.  It's already kinda expected that Web Inspector is overriding the page's behavior for these kinds of things, so `Forced` is a bit redundant.

dcrousso commented on Source/JavaScriptCore/inspector/protocol/Page.json:
```diff
@@ -385,6 +397,14 @@
             ]
+        },
+        {
+            "name": "accessibilitySettingsDidChange",
```
Why limit this to accessibility?  Couldn't we include something like `prefers-reduced-data` in here?  What about a more general `userPreferencesDidChange` instead?

dcrousso commented on Source/WebCore/inspector/agents/InspectorPageAgent.cpp:
```diff
@@ -890,6 +905,57 @@ void InspectorPageAgent::frameClearedScheduledNavigation(Frame& frame)
+    // An accessibility setting is considered `forced` if its current forced state is the opposite of the system value, whichever that may be.
+    bool forced = false;
+
+    if (!settingEnabledOnSystem && forcedAccessibilityValue == ForcedAccessibilityValue::On) {
```
I think we can massively simplify this
```
bool enabled = forcedAccessibilityValue == ForcedAccessibilityValue::On || (settingEnabledOnSystem && forcedAccessibilityValue == ForcedAccessibilityValue::System);
bool forced = (settingEnabledOnSystem && forcedAccessibilityValue == ForcedAccessibilityValue::Off) || (!settingEnabledOnSystem && forcedAccessibilityValue == ForcedAccessibilityValue::On);
```

dcrousso commented on Source/WebCore/page/Page.cpp:
```diff
@@ -3324,13 +3324,21 @@ void Page::setCaptionUserPreferencesStyleSheet(const String& styleSheet)
-void Page::accessibilitySettingsDidChange()
+void Page::accessibilitySettingsDidChange(bool forced)
 {
+    if (!forced) {
```
I don't think this will work as we want it to.  What if the developer

1. overrides in Web Inspector
2. changes the system (which would remove the Web Inspector override0
3. changes the system again

we should *always* have whatever the developer has set as the current state, no matter what the system does underneath them (or what they do to the system)

dcrousso commented on Source/WebInspectorUI/UserInterface/Controllers/CSSManager.js:
```diff
@@ -296,6 +298,10 @@ WI.CSSManager = class CSSManager extends WI.Object
 
+    get prefersReducedMotion() { return this._prefersReducedMotion; }
+
+    get prefersContrast() { return this._prefersContrast; }
```
since these are both simple getters, you can move these to be right under `get propertyNameCompletions`, each on a single line

dcrousso commented on Source/WebInspectorUI/UserInterface/Controllers/CSSManager.js:
```diff
@@ -531,6 +537,26 @@ WI.CSSManager = class CSSManager extends WI.Object
 
+    accessibilitySettingsDidChange(accessibilitySettings)
+    {
+        console.assert(Array.isArray(accessibilitySettings), accessibilitySettings);
```
this is unnecessary since this is always called from the protocol

dcrousso commented on Source/WebInspectorUI/UserInterface/Controllers/CSSManager.js:
```diff
@@ -531,6 +537,26 @@ WI.CSSManager = class CSSManager extends WI.Object
+                this._prefersContrast = {enabled, forced};
+                break;
+            default:
+                console.error("Unexpected protocol AccessibilitySetting name:", name);
```
please use `WI.reportInternalError`

also this is missing a `break;`

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/DOMTreeContentView.js:
```diff
@@ -58,6 +58,12 @@ WI.DOMTreeContentView = class DOMTreeContentView extends WI.ContentView
         this._forceAppearanceButtonNavigationItem.visibilityPriority = WI.NavigationItem.VisibilityPriority.Low;
         this._forceAppearanceButtonNavigationItem.enabled = WI.cssManager.canForceAppearance();
 
+        // COMPATIBILITY (macOS 13.3, iOS 16.4): `ForcedPrefersReducedMotion` value for `Page.Setting` did not exist yet.
```
this should be `macOS 13.0, iOS 16.0` as these comments correspond to the last version that did not contain the relevant protocol stuff

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/DOMTreeContentView.js:
```diff
@@ -58,6 +58,12 @@ WI.DOMTreeContentView = class DOMTreeContentView extends WI.ContentView
         this._forceAppearanceButtonNavigationItem.enabled = WI.cssManager.canForceAppearance();
 
+        // COMPATIBILITY (macOS 13.3, iOS 16.4): `ForcedPrefersReducedMotion` value for `Page.Setting` did not exist yet.
+        this._toggleAccessibilitySettingsNavigationItem = new WI.ActivateButtonNavigationItem("toggle-accessibility-settings", "Toggle accessibility settings", "Toggled accessibility settings differ from system settings", "Images/Accessibility.svg", 16, 16);
```
it's odd for the tooltip to go from an action (e.g. "do this") to a statement (e.g. "this is XYZ")

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/DOMTreeContentView.js:
```diff
@@ -694,6 +702,77 @@ WI.DOMTreeContentView = class DOMTreeContentView extends WI.ContentView
 
+    _handleAccessibilitySettingsButtonClicked()
+    {
+        if (!this._toggleAccessibilitySettingsNavigationItem.enabled)
```
this shouldn't be possible

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/DOMTreeContentView.js:
```diff
@@ -694,6 +702,77 @@ WI.DOMTreeContentView = class DOMTreeContentView extends WI.ContentView
+            return;
+        }
+
+        const createCheckbox = (container, label, setting, value) => {
```
Style: we only use `const` for things that don't change between Web Inspector sessions (i.e. true constants).  please use `let`

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/DOMTreeContentView.js:
```diff
@@ -694,6 +702,77 @@ WI.DOMTreeContentView = class DOMTreeContentView extends WI.ContentView
+            return checkboxElement;
+        }
+
+        const calculateTargetFrame = () => {
```
ditto (:716)

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/DOMTreeContentView.js:
```diff
@@ -694,6 +702,77 @@ WI.DOMTreeContentView = class DOMTreeContentView extends WI.ContentView
+            checkboxElement.type = "checkbox";
+            checkboxElement.checked = value;
+            checkboxElement.addEventListener("change", (event) => {
+                this._setAccessibilitySettingForAllTargets(setting, event.target.checked);
```
NIT: i think `checkboxElement` would be clearer

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/DOMTreeContentView.js:
```diff
@@ -694,6 +702,77 @@ WI.DOMTreeContentView = class DOMTreeContentView extends WI.ContentView
+            if (!setting)
+                return;
+
+            let labelElement = container.appendChild(document.createElement("label"));
```
the `container` is the same value every time this method is called.  could we move `createCheckbox` down below `let contentElement = ...` and then just use it directly instead?

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/DOMTreeContentView.js:
```diff
@@ -694,6 +702,77 @@ WI.DOMTreeContentView = class DOMTreeContentView extends WI.ContentView
+        let contentElement = document.createElement("div");
+        contentElement.classList.add("accessibility-settings-content");
+
+        // COMPATIBILITY (macOS 13.3, iOS 16.4): `ForcedPrefersReducedMotion` value for `Page.Setting` did not exist yet.
```
ditto (:61)

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/DOMTreeContentView.js:
```diff
@@ -694,6 +702,77 @@ WI.DOMTreeContentView = class DOMTreeContentView extends WI.ContentView
+        if (InspectorBackend.Enum.Page?.Setting?.ForcedPrefersReducedMotion)
+            this._prefersReducedMotionCheckbox = createCheckbox(contentElement, "Reduce motion", InspectorBackend.Enum.Page.Setting.ForcedPrefersReducedMotion, WI.cssManager.prefersReducedMotion.enabled);
+
+        // COMPATIBILITY (macOS 13.3, iOS 16.4): `ForcedPrefersContrast` value for `Page.Setting` did not exist yet.
```
ditto (:61)

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/DOMTreeContentView.js:
```diff
@@ -694,6 +702,77 @@ WI.DOMTreeContentView = class DOMTreeContentView extends WI.ContentView
+        this._accessibilitySettingsPopover.presentNewContentWithFrame(contentElement, calculateTargetFrame(), preferredEdges);
+    }
+
+    _setAccessibilitySettingForAllTargets(setting, value)
```
please move this to `WI.CSSManager` as we prefer to keep protocol code out of view stuff

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/DOMTreeContentView.js:
```diff
@@ -694,6 +702,77 @@ WI.DOMTreeContentView = class DOMTreeContentView extends WI.ContentView
+        this._prefersContrastCheckbox.checked = WI.cssManager.prefersContrast.enabled;
+
+        // Highlight the icon whenever forced preferences differ from default system preferences.
+        this._toggleAccessibilitySettingsNavigationItem.activated = [WI.cssManager.prefersReducedMotion, WI.cssManager.prefersContrast].some(pref => pref.forced === true);
```
Style: missing `(` `)` around arrow function parameter(s)
Style: the `=== true` is redundant

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/DOMTreeContentView.js:
```diff
@@ -694,6 +702,77 @@ WI.DOMTreeContentView = class DOMTreeContentView extends WI.ContentView
+        this._prefersReducedMotionCheckbox.checked = WI.cssManager.prefersReducedMotion.enabled;
+        this._prefersContrastCheckbox.checked = WI.cssManager.prefersContrast.enabled;
+
+        // Highlight the icon whenever forced preferences differ from default system preferences.
```
id remove this comment as it's kinda redundant if you read the code

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/Main.css:
```diff
@@ -579,6 +578,11 @@ body[dir=rtl] [dir=ltr] .go-to-arrow {
 
+.accessibility-settings-content {
+    display: flex;
+    flex-flow: column;
```
NIT: id prefer to use `flex-direction` if you're not setting a `flex-wrap` value

dcrousso commented on Source/JavaScriptCore/inspector/protocol/Page.json:
```diff
@@ -24,6 +24,21 @@
+        {
+            "id": "UserPreferenceValue",
+            "type": "string",
+            "enum": ["System", "On", "Off"],
```
I don't think we need an explicit `"System"` value since the use of this is already `"optional": true`.  The "revert to system" should be (and looking at the code it seems like it already is) just not providing the `value`.

Also, I don't think `"On"` and `"Off"` are really the right values for this.  I think was should just make the `value` into a `"type": "string"` since each one of these user preferences (and the others we should add in the future) could have arbitrary values that are not `"On"` and `"Off"`.  Each `UserPreference` should just do it's own parsing in the backend to figure out what each possible string value is allowed.

dcrousso commented on Source/WebInspectorUI/UserInterface/Controllers/CSSManager.js:
```diff
@@ -389,6 +389,14 @@ WI.CSSManager = class CSSManager extends WI.Object
+    overrideUserPreference(preference, value)
+    {
+        for (let target of WI.targets) {
+            if (target.hasCommand("Page.overrideUserPreference"))
```
this needs a `// COMPATIBILITY (macOS 13.0, iOS 16.0): ...`

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/DOMTreeContentView.js:
```diff
@@ -694,6 +702,97 @@ WI.DOMTreeContentView = class DOMTreeContentView extends WI.ContentView
+            });
+
+            return selectElement;
+        }
```
Style: missing `;`

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/DOMTreeContentView.js:
```diff
@@ -694,6 +702,97 @@ WI.DOMTreeContentView = class DOMTreeContentView extends WI.ContentView
+
+        let calculateTargetFrame = () => {
+            return WI.Rect.rectFromClientRect(this._overrideUserPreferencesNavigationItem.element.getBoundingClientRect()).pad(2);
+        }
```
ditto (:752)

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/DOMTreeContentView.js:
```diff
@@ -694,6 +702,97 @@ WI.DOMTreeContentView = class DOMTreeContentView extends WI.ContentView
+            return WI.Rect.rectFromClientRect(this._overrideUserPreferencesNavigationItem.element.getBoundingClientRect()).pad(2);
+        }
+
+        let preferredEdges = [WI.RectEdge.MAX_Y, WI.RectEdge.MAX_X];
```
Style: `const`

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/DOMTreeContentView.js:
```diff
@@ -694,6 +702,97 @@ WI.DOMTreeContentView = class DOMTreeContentView extends WI.ContentView
+        let contentElement = document.createElement("div");
+        contentElement.classList.add("user-preferences-content");
+
+        let createSelectElement = (label, preference, values = [], defaultValue) => {
```
Style: no space between `[` `]`

also, it looks like every caller of this provides values for that (and we should expect that) so I don't think we even need the optional fallback or the `!values.length` check

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/DOMTreeContentView.js:
```diff
@@ -694,6 +702,97 @@ WI.DOMTreeContentView = class DOMTreeContentView extends WI.ContentView
+        let preferredEdges = [WI.RectEdge.MAX_Y, WI.RectEdge.MAX_X];
+
+        this._userPreferencesOverridesPopover = new WI.Popover(this);
+        this._userPreferencesOverridesPopover.windowResizeHandler = () => {
```
Aside: we should probably audit all `set windowResizeHandler` to make sure that we're not keeping things alive longer than they should be (e.g. `this` is being kept alive here)

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/DOMTreeContentView.js:
```diff
@@ -694,6 +702,97 @@ WI.DOMTreeContentView = class DOMTreeContentView extends WI.ContentView
+        let accessibilityHeader = contentElement.appendChild(document.createElement("h5"));
+        accessibilityHeader.textContent = WI.UIString("Accessibility Preferences", "Accessibility Preferences @ User Preferences Overrides", "Header for section with accessibility user preferences.");
+
+        let userPreferenceValues = [InspectorBackend.Enum.Page.UserPreferenceValue.System, InspectorBackend.Enum.Page.UserPreferenceValue.On, InspectorBackend.Enum.Page.UserPreferenceValue.Off];
```
ditto (:758)

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/DOMTreeContentView.js:
```diff
@@ -694,6 +702,97 @@ WI.DOMTreeContentView = class DOMTreeContentView extends WI.ContentView
+
+        let userPreferenceValues = [InspectorBackend.Enum.Page.UserPreferenceValue.System, InspectorBackend.Enum.Page.UserPreferenceValue.On, InspectorBackend.Enum.Page.UserPreferenceValue.Off];
+
+        // COMPATIBILITY (macOS 13.0, iOS 16.0): `PrefersReducedMotion` value for `Page.UserPreference` did not exist yet.
```
we should be able to assume this since we can only reach this code if this is already true (i.e. `_overrideUserPreferencesNavigationItem` is only clickable if the protocol supports this)

please either remove these checks or if you really wanna be safe add some sort of check on :777 that bails if the `contentElement` is empty

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/Main.css:
```diff
@@ -563,14 +563,13 @@ body[dir=rtl] [dir=ltr] .go-to-arrow {
 }
 
-.device-settings-content label + label {
+:is(.device-settings-content, .user-preferences-content) label + label {
```
please move all `.user-preferences-content` styles to `Source/WebInspectorUI/UserInterface/Views/DOMTreeContentView.css` as that's where the corresponding DOM is created

i.e. the styles for `.device-settings-content` exist here because the corresponding DOM is created in `Source/WebInspectorUI/UserInterface/Base/Main.js`

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/Main.css:
```diff
@@ -579,6 +578,27 @@ body[dir=rtl] [dir=ltr] .go-to-arrow {
+
+.user-preferences-content > h5 {
+    margin-block-start: 4px;
+    margin-block-end: 8px;
```
`margin-inline: 4px 8px;`

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/Main.css:
```diff
@@ -579,6 +578,27 @@ body[dir=rtl] [dir=ltr] .go-to-arrow {
+
+.user-preferences-content label {
+    display: flex;
+    flex-direction: row;
```
NIT: this is unnecessary since it's the default

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/Main.css:
```diff
@@ -579,6 +578,27 @@ body[dir=rtl] [dir=ltr] .go-to-arrow {
+}
+
+.user-preferences-content label > span {
+    flex: auto;
```
NIT: is this necessary?

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/Main.css:
```diff
@@ -579,6 +578,27 @@ body[dir=rtl] [dir=ltr] .go-to-arrow {
+    font-size: 11px;
+}
+
+.user-preferences-content label {
```
Isn't the `label` a direct child of `.user-preferences-content`?  If so, please use `.user-preferences-content > label`.

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/Main.css:
```diff
@@ -579,6 +578,27 @@ body[dir=rtl] [dir=ltr] .go-to-arrow {
+    flex-direction: row;
+}
+
+.user-preferences-content label > span {
```
ditto (:593)

patrickangle commented on Source/JavaScriptCore/inspector/protocol/Page.json:
```diff
@@ -24,6 +24,27 @@
             ]
         },
+        {
+            "id": "UserPreference",
```
Nit: I'd move `UserPreference` down below the other enums at the top of this file.

patrickangle commented on Source/JavaScriptCore/inspector/protocol/Page.json:
```diff
@@ -385,6 +415,14 @@
+            "description": "Fired when the default value of a user preference changes at the system level.",
+            "targetTypes": ["page"],
+            "parameters": [
+                { "name": "preferences", "type": "array", "items": { "$ref": "UserPreference" }, "description": "List of user preferences that can be overriden and their default (system) values." }
```
"and their default" -> "and their new system (default)" to also communicate the change in terms of why it is occurring?

rcaliman-apple replied:
Done.

patrickangle commented on Source/WebCore/inspector/agents/InspectorPageAgent.cpp:
```diff
@@ -45,6 +45,7 @@
 #include "FrameLoader.h"
 #include "FrameSnapshotting.h"
 #include "FrameView.h"
+#include "ForcedAccessibilityValue.h"
```
Nit: alphabetize includes

rcaliman-apple replied:
Done

patrickangle reviewed (commented):
Backend looks good with a few nits, I'll look at frontend after lunch.

patrickangle commented on Source/WebInspectorUI/UserInterface/Controllers/CSSManager.js:
```diff
@@ -389,6 +403,23 @@ WI.CSSManager = class CSSManager extends WI.Object
+                else {
+                    target.PageAgent.overrideUserPreference(preference); // Pass no value to remove the override and restore the system default.
+                    this._overridenUserPreferences.delete(preference);
+                }
```
I think this can be simplified a bit:
```
target.PageAgent.overrideUserPreference(preference, value);
if (value)
    this._overridenUserPreferences.set(preference, value);
else
    this._overridenUserPreferences.delete(preference);
```

rcaliman-apple replied:
Good catch.

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/DOMTreeContentView.css:
```diff
@@ -85,3 +85,27 @@ body:not(.window-inactive, .window-docked-inactive) .content-view.dom-tree .tree
+    grid-template-columns: auto auto;
+}
+
+.user-preferences-content > h5 {
```
Why not just use `h1` here and style it appropriately, since this is the top-level type of header inside this popover?

rcaliman-apple replied:
Done.

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/DOMTreeContentView.css:
```diff
@@ -85,3 +85,27 @@ body:not(.window-inactive, .window-docked-inactive) .content-view.dom-tree .tree
     display: none;
 }
+
+.user-preferences-content {
```
Can we use a more specific selector for all of these these? At the very least include the popover class?

rcaliman-apple replied:
Moved to a new file `OverrideUserPreferencesPopover.css` and made it slightly more specific with `.popover` prefix.

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/DOMTreeContentView.js:
```diff
@@ -694,6 +702,108 @@ WI.DOMTreeContentView = class DOMTreeContentView extends WI.ContentView
+            return WI.UIString("Off", "Off @ User Preferences Overrides", "Label for a preference that is turned off.");
+        }
+
+        console.assert(false, "Unknown user preference");
```
Nit: Add the `preferenceValue` as a third param to help debugging later if we end up here.

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/DOMTreeContentView.js:
```diff
@@ -694,6 +702,108 @@ WI.DOMTreeContentView = class DOMTreeContentView extends WI.ContentView
+            return;
+
+        for (let [preferenceName, systemValue] of WI.cssManager.defaultUserPreferences) {
+            let defaultOptionElement = this._userPreferencesOverridesPopover.element.querySelector(`select[name=${preferenceName}] > option[value=${WI.CSSManager.UserPreferenceDefaultValue}]`);
```
I'm skeptical of this approach to finding the select elements to update them... Could we consider instead of a subclass of `WI.Popover` (an example being `WI.CSSDocumentationPopover`) that keeps a more structured understand of the elements for the preferences it is showing?

rcaliman-apple replied:
Moved to a new `OverrideUserPreferencesPopover` that subclasses `Popover` and holding an internal map of elements to change.

patrickangle reviewed (commented):
This is really close - my two main concerns are the lack of any tests and my comment about creating a popover subclass so we don't have to query for elements (which has the added benefit of not growing DOMTreeContentView as we add more preferences).

For testing, I'd suggest a few test cases that define unique styles for the Yes/No states of these two preferences, and then applies the preference and refreshes the applied styles, ensuring that the correct style is being applied to the element.

patrickangle commented on LayoutTests/inspector/page/overrideUserPreference.html:
```diff
@@ -0,0 +1,131 @@
+    }
+    </style>
+</body>
+</html>
```
Style: needs a newline at the end of the file

rcaliman-apple replied:
Done

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/OverrideUserPreferencesPopover.css:
```diff
@@ -0,0 +1,48 @@
+
+.popover .user-preferences-content > select {
+    grid-column: 2;
+}
```
Style: needs a newline at the end of the file

rcaliman-apple replied:
Done

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/OverrideUserPreferencesPopover.js:
```diff
@@ -0,0 +1,150 @@
+
+        this._targetElement = null;
+        this._defaultValueOptionElementsMap = new Map;
+        this.windowResizeHandler = this._presentOverTargetElement.bind(this);
```
Aside: It's a shame every subclass requires this.

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/DOMTreeContentView.js:
```diff
@@ -694,6 +702,23 @@ WI.DOMTreeContentView = class DOMTreeContentView extends WI.ContentView
+    {
+        if (this._userPreferencesOverridesPopover) {
+            this._userPreferencesOverridesPopover.dismiss();
+            this._userPreferencesOverridesPopover = null;
```
Won't this be done by the delegate function `didDismissPopover(popover)`?

rcaliman-apple replied:
~~Yes, `null`-ing the popover instance happens in `didDismissPopover()` but that occurs after a CSS `transitionend` event that takes a few hundred milliseconds to trigger. See [Popover.js?L189-200](https://sourcegraph.com/github.com/WebKit/WebKit@f7c30648762020ca176afcd3932b878a07b44994/-/blob/Source/WebInspectorUI/UserInterface/Views/Popover.js?L189-200)~~

~~`this._handleOverrideUserPreferencesButtonClicked()` reacts to clicks on the trigger element. There's a chance the user clicks repeatedly and very quickly. `null`-ing immediately here saves the trouble with potential zombie instances living long enough to cause issues.~~

Nevermind. You're right. I misunderstood how `Popover.dismiss()` works. A subsequent click on the trigger is a click outside the popover, therefore the popover will go through the dismiss flow automatically. Me calling it explicitly triggered a duplicate  which tripped assertions when removing event handlers that were already removed. That's how I learned what you meant :)

I removed the manual dismissal and `null`-ification here in favor of `didDismissPopover(popover)`

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/OverrideUserPreferencesPopover.css:
```diff
@@ -0,0 +1,48 @@
+}
+
+.popover .user-preferences-content > h1 {
+    grid-column: 1 / 3;
```
We can actually express this as `grid-column: 1 / -1` to span the width of the grid, since negative line numbers count from right to left (in left-to-right writing modes). This way if the grid ever got wider, the title would still span the full width of the grid without a change.

rcaliman-apple replied:
Good idea!

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/OverrideUserPreferencesPopover.css:
```diff
@@ -0,0 +1,48 @@
+.popover .user-preferences-content > label {
+    grid-column: 1;
+    align-self: center;
+    margin-inline-end: 4px;
```
Do we need both this and the gap? Could we just set `gap: 4px 8px` (vertical gap followed by horizontal gap) in `.popover .user-preferences-content` instead?

rcaliman-apple replied:
Yes, that works too. Done.

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/OverrideUserPreferencesPopover.js:
```diff
@@ -0,0 +1,150 @@
+
+        const userPreferenceValues = [InspectorBackend.Enum.Page.UserPreferenceValue.On, InspectorBackend.Enum.Page.UserPreferenceValue.Off];
+
+        // COMPATIBILITY (macOS 13.0, iOS 16.0): `PrefersReducedMotion` value for `Page.UserPreferenceName` did not exist yet.
```
Since we should never reach here when inspecting a target running an older OS, I don't think we need these compatibility checks. But maybe we do want to assert in the constructor that `InspectorBackend.hasCommand("Page.overrideUserPreference")` is true, or a similar expression (with a compat comment as well). Then we can safely assume in this entire file that the stuff you are adding in the patch exists.

rcaliman-apple replied:
I'll keep this as-is and not guard this whole component behind `InspectorBackend.hasCommand("Page.overrideUserPreference")` because I have a follow-up patch which implements the backwards-compatible method to toggle appearance with the same UI.

patrickangle commented on LayoutTests/inspector/page/overrideUserPreference.html:
```diff
@@ -0,0 +1,131 @@
+
+    @media (prefers-reduced-motion) {
+        body {
+            --motion: on;
```
Nit: strictly speaking this would be turning motion off. Maybe `--test-prefers-reduced-motion: on;`? (Ditto for `--test-prefers-contrast` as well for consistency.)

rcaliman-apple replied:
Indeed. This makes it clearer.

patrickangle reviewed (approved):
rs=me with a few remaining style/nits

dcrousso commented on Source/JavaScriptCore/inspector/protocol/Page.json:
```diff
@@ -24,6 +24,27 @@
+        {
+            "id": "UserPreferenceValue",
+            "type": "string",
+            "enum": ["Off", "On"],
```
I thought the plan was to have this be a `"type": "any"` since each user presence could be something that could have different values (e.g. `prefers-contrast: more` vs `prefers-reduced-motion: reduce` vs etc.).

rcaliman-apple replied:
I don't understand how `"type": "any"` makes it any different.

For the upcoming patch that implements overriding color scheme (appearance), the protocol types are augmented like this:


```diff
        {
             "id": "UserPreferenceName",
             "type": "string",
-            "enum": ["PrefersReducedMotion", "PrefersContrast"],
+            "enum": ["PrefersReducedMotion", "PrefersContrast", "PrefersColorScheme"],
             "description": "User preference name."
         },
                 {
            "id": "UserPreferenceValue",
            "type": "string",
-            "enum": ["Off", "On"],
+            "enum": ["Off", "On", "Light", "Dark"],
            "description": "User preference value."
        },
```

The implementation of `InspectorPageAgent::overrideUserPreference(Protocol::Page::UserPreferenceName, std::optional<Protocol::Page::UserPreferenceValue>&&)` explicitly checks the preference name against expected values to set the overrides, ignoring the unexpected.

With `"type": "any"` wouldn't we do the same on the backend?

dcrousso replied:
i suppose yeah it would be nice to have a way for the frontend to check for value support (in addition to the name support you're adding here)

in that case, we should change `"On"` and `"Off"` to be the actual values used by the relevant name(s) (e.g. `"no-preference"`, `"reduce"`, `"less"`, `"more"`, etc.)

rcaliman-apple replied:
Ok. I added explicit preferences values: "NoPreference", "More" and "Reduce". 
They're handled on the backend corresponding to their preference name.

dcrousso commented on Source/WebInspectorUI/UserInterface/Controllers/CSSManager.js:
```diff
@@ -306,6 +307,19 @@ WI.CSSManager = class CSSManager extends WI.Object
+    set defaultUserPreferences(userPreferences)
+    {
+        for (let userPreference of userPreferences)
+            this.defaultUserPreferences.set(userPreference.name, userPreference.value)
```
Should we also do a `this._defaultUserPreferences.clear()` at the start of this function?

Style: please use `_defaultUserPreferences` since we're inside the class that declares the underlying member variable

rcaliman-apple replied:
Removed the setter per other comment. 

I won't clear the `this._defaultUserPreferences` on this call yet. I have a follow-up patch that's implementing a backwards compatible appearance override. The handler for the old `defaultAppearanceDidChange` event will add its value to `this._defaultUserPreferences`. A subsequent clear would drop it because the `defaultUserPreferencesDidChange` won't have the value for the legacy default appearance.

dcrousso replied:
we probably should have at least _some_ way of clearing values from `_defaultUserPreferences` 🤔

dcrousso commented on Source/WebInspectorUI/UserInterface/Controllers/CSSManager.js:
```diff
@@ -306,6 +307,19 @@ WI.CSSManager = class CSSManager extends WI.Object
+            this.defaultUserPreferences.set(userPreference.name, userPreference.value)
+    }
+
+    get overridenUserPreferences() { return this._overridenUserPreferences; }
```
Style: single line `get` should be at the top of the `// Public` section (after any other single line `get`)

rcaliman-apple replied:
Done.

dcrousso commented on Source/WebInspectorUI/UserInterface/Controllers/CSSManager.js:
```diff
@@ -389,6 +403,23 @@ WI.CSSManager = class CSSManager extends WI.Object
+    {
+        for (let target of WI.targets) {
+            // COMPATIBILITY (macOS 13.0, iOS 16.0): `Page.overrideUserPreference()` did not exist yet.
+            if (target.hasCommand("Page.overrideUserPreference")) {
```
NIT: Can we make this into a `continue` instead of having to indent everything?

rcaliman-apple replied:
I'll leave this as-is for now. A follow-up patch implementing a backwards compatible way to override color-scheme/appearance will add another condition here to use the legacy protocol method. `continue` would miss it.

dcrousso commented on Source/WebInspectorUI/UserInterface/Controllers/CSSManager.js:
```diff
@@ -531,6 +562,13 @@ WI.CSSManager = class CSSManager extends WI.Object
 
+    defaultUserPreferencesDidChange(userPreferences)
+    {
+        this.defaultUserPreferences = userPreferences;
```
This is probably something we never want anyone else to be able to set, so I think we should remove the `set defaultUserPreferences` and instead move that code to be inlined here.

rcaliman-apple replied:
Yes, that's true. I removed the setter.

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/DOMTreeContentView.js:
```diff
@@ -58,6 +58,12 @@ WI.DOMTreeContentView = class DOMTreeContentView extends WI.ContentView
         this._forceAppearanceButtonNavigationItem.enabled = WI.cssManager.canForceAppearance();
 
+        // COMPATIBILITY (macOS 13.0, iOS 16.0): `Page.overrideUserPreference` did not exist yet.
+        this._overrideUserPreferencesNavigationItem = new WI.ActivateButtonNavigationItem("toggle-user-preferences", WI.UIString("Override user preferences"), WI.UIString("User preferences overriden"), "Images/AppearanceOverride.svg", 16, 16);
```
Now that the Device Settings navigation item is always shown (i.e. visible on macOS), I think we should actually move the popover UI to be inside that, so that the user preference settings can be accessed from any tab (e.g. something like `prefers-reduced-data` would really be more of a concern for the Network Tab, so having to switch between the Elements Tab and Network Tab to do anything with it would be very annoying).

patrickangle replied:
I disagree with adding these setting to the "Device Settings" panel. There is a key distinction between those settings and these setting in that the settings Razvan is adding our valid web-exposed states that the user controls, where the controls in the "Device Settings" panel are, generally (yes JS is the exception in Safari via the Security settings panel), a state that users will never be in.

I understand the desire to future-proof the position of these toggles in that way, but even `prefers-reduced-data` is arguably still about CSS first, and networking second, because that query does not itself change anything about the network requests, its still up to CSS to define different content to be loaded. Ultimately these are settings (and potential future settings) that change what CSS rules are applied on a page.

Also, the device settings coming to a local session is still being massaged after some internal debate, so isn't presently landed/available anyways.

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/OverrideUserPreferencesPopover.js:
```diff
@@ -0,0 +1,150 @@
+        this._targetElement = targetElement;
+        this.content = this._createContentElement();
+
+        WI.cssManager.addEventListener(WI.CSSManager.Event.DefaultUserPreferencesDidChange, this._defaultUserPreferencesDidChange, this);
```
we should also `removeEventListener` inside `dismiss`

rcaliman-apple replied:
Good catch!

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/OverrideUserPreferencesPopover.js:
```diff
@@ -0,0 +1,150 @@
+
+        let labelElement = contentElement.appendChild(document.createElement("label"));
+        labelElement.textContent = label;
+        labelElement.setAttribute("for", preferenceName);
```
NIT: It's a bit odd that we use a protocol value in the DOM.  Can we have this function take an `id` as the first argument and use that instead?

rcaliman-apple replied:
Added an `id` parameter.

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/OverrideUserPreferencesPopover.js:
```diff
@@ -0,0 +1,150 @@
+            if (value === selectedValue)
+                optionElement.setAttribute("selected", true);
+
+            if (value === defaultValue)
```
I think we should also add a `<hr>` after the "System (...)" `<option>` so that it's even clearer that it's not one of the actual values (this is what we do for the Appearance dropdown in the Settings Tab).

rcaliman-apple replied:
Done

rcaliman-apple committed:
Web Inspector: Add emulation toggles for prefers-reduced-motion and prefers-contrast
https://bugs.webkit.org/show_bug.cgi?id=246882

Reviewed by Patrick Angle.

Adds controls in Web Inspector to override preferences
for reduced motion and increased contrast for the inspected page.

These correspond to CSS media features `prefers-reduced-motion: reduce | no-preference`
and `prefers-contrast: more | no-preference`.

The UI in Web Inspector starts off with preference values set to `System`
meaning there is no override applied. A user can pick `On` or `Off` to
force a value for the preference on the inspected page for the duration of
the Web Inspector session.

When a preference is overriden in Web Inspector, changes after the fact to
corresponding settings in System Settings > Accessibility > Display
have no effect on the inspected page until Web Inspector is closed or
the preference values are returned to `System`.

* LayoutTests/inspector/page/overrideUserPreference-expected.txt: Added.
* LayoutTests/inspector/page/overrideUserPreference.html: Added.
* Source/JavaScriptCore/inspector/protocol/Page.json:
* Source/WebCore/inspector/InspectorInstrumentation.cpp:
(WebCore::InspectorInstrumentation::accessibilitySettingsDidChangeImpl):
* Source/WebCore/inspector/InspectorInstrumentation.h:
(WebCore::InspectorInstrumentation::accessibilitySettingsDidChange):
* Source/WebCore/inspector/agents/InspectorPageAgent.cpp:
(WebCore::InspectorPageAgent::enable):
(WebCore::InspectorPageAgent::disable):
(WebCore::InspectorPageAgent::overrideUserPreference):
(WebCore::InspectorPageAgent::overridePrefersReducedMotion):
(WebCore::InspectorPageAgent::overridePrefersContrast):
(WebCore::InspectorPageAgent::accessibilitySettingsDidChange):
(WebCore::InspectorPageAgent::defaultUserPreferencesDidChange):
* Source/WebCore/inspector/agents/InspectorPageAgent.h:
* Source/WebCore/page/Page.cpp:
(WebCore::Page::accessibilitySettingsDidChange):
* Source/WebInspectorUI/Localizations/en.lproj/localizedStrings.js:
* Source/WebInspectorUI/UserInterface/Controllers/CSSManager.js:
(WI.CSSManager):
(WI.CSSManager.prototype.get overridenUserPreferences):
(WI.CSSManager.prototype.get defaultUserPreferences):
(WI.CSSManager.prototype.overrideUserPreference):
(WI.CSSManager.prototype.defaultUserPreferencesDidChange):
* Source/WebInspectorUI/UserInterface/Images/AppearanceOverride.svg: Added.
* Source/WebInspectorUI/UserInterface/Main.html:
* Source/WebInspectorUI/UserInterface/Protocol/PageObserver.js:
(WI.PageObserver.prototype.defaultUserPreferencesDidChange):
* Source/WebInspectorUI/UserInterface/Views/DOMTreeContentView.js:
(WI.DOMTreeContentView):
(WI.DOMTreeContentView.prototype.get navigationItems):
(WI.DOMTreeContentView.prototype._handleOverrideUserPreferencesButtonClicked):
(WI.DOMTreeContentView.prototype._overridenUserPreferencesDidChange):
(WI.DOMTreeContentView.prototype.didDismissPopover):
* Source/WebInspectorUI/UserInterface/Views/OverrideUserPreferencesPopover.css: Added.
(.popover .user-preferences-content):
(.popover .user-preferences-content > h1):
(.popover .user-preferences-content > label):
(.popover .user-preferences-content > select):
* Source/WebInspectorUI/UserInterface/Views/OverrideUserPreferencesPopover.js: Added.
(WI.OverrideUserPreferencesPopover):
(WI.OverrideUserPreferencesPopover.prototype.show):
(WI.OverrideUserPreferencesPopover.prototype.dismiss):
(WI.OverrideUserPreferencesPopover.prototype._presentOverTargetElement):
(WI.OverrideUserPreferencesPopover.prototype._createSelectElement):
(WI.OverrideUserPreferencesPopover.prototype._createContentElement):
(WI.OverrideUserPreferencesPopover.prototype._userPreferenceValueLabel):
(WI.OverrideUserPreferencesPopover.prototype._defaultUserPreferencesDidChange):

Canonical link: https://commits.webkit.org/257197@main
